core/src/main/scala/cats/data/Streaming.scala

```diff
@@ -0,0 +1,953 @@
+    loop(Empty(), as.length - 1)
+  }
+
+  /**
+   * Create a stream from a list.
+   *
+   * The stream will be eagerly evaluated.
+   */
+  def fromList[A](as: List[A]): Streaming[A] = {
+    def loop(s: Streaming[A], ras: List[A]): Streaming[A] =
+      ras match {
+        case Nil => s
+        case a :: rt => loop(This(a, Now(s)), rt)
+      }
+    loop(Empty(), as.reverse)
+  }
+
+  /**
+   * Create a stream from an iterable.
+   *
+   * The stream will be eagerly evaluated.
+   */
+  def fromIterable[A](as: Iterable[A]): Streaming[A] =
+    fromIteratorUnsafe(as.iterator)
+
+  /**
+   * Create a stream from an iterator.
+   *
+   * The stream will be created lazily, to support potentially large
+   * (or infinite) iterators. Iterators passed to this method should
+   * not be used elsewhere -- doing so will result in problems.
+   *
+   * The use case for this method is code like .fromIterable, which
+   * creates an iterator for the express purpose of calling this
+   * method.
+   */
+  def fromIteratorUnsafe[A](it: Iterator[A]): Streaming[A] =
+    if (it.hasNext) This(it.next, Later(fromIteratorUnsafe(it))) else Empty()
+
+  /**
+   * Create a self-referential stream.
+   */
+  def knot[A](f: Eval[Streaming[A]] => Streaming[A], memo: Boolean = false): Streaming[A] = {
+    lazy val s: Eval[Streaming[A]] = if (memo) Later(f(s)) else Always(f(s))
+    s.value
+  }
+
+  /**
+   * Continually return a constant value.
+   */
+  def continually[A](a: A): Streaming[A] =
+    knot(s => This(a, s), memo = true)
+
+  /**
+   * Continually return the result of a thunk.
+   *
+   * This method only differs from `continually` in that the thunk may
+   * not be pure. For this reason (and unlike continually), this
+   * stream is memoized to ensure that repeated traversals produce the
+   * same results.
+   */
+  def thunk[A](f: () => A): Streaming[A] =
+    knot(s => This(f(), s), memo = true)
+
+  /**
+   * Produce an infinite stream of values given an initial value and a
+   * tranformation function.
+   */
+  def infinite[A](a: A)(f: A => A): Streaming[A] =
+    This(a, Always(infinite(f(a))(f)))
+
+  /**
+   * Stream of integers starting at n.
+   */
+  def from(n: Int): Streaming[Int] =
+    infinite(n)(_ + 1)
+
+  /**
+   * Provide a stream of integers starting with `start` and ending
+   * with `end` (i.e. inclusive).
+   */
+  def interval(start: Int, end: Int): Streaming[Int] =
+    if (start > end) Empty() else This(start, Always(interval(start + 1, end)))
+
+  /**
+   * Produce a stream given an "unfolding" function.
+   *
+   * None represents an empty stream. Some(a) reprsents an initial
+   * element, and we can compute the tail (if any) via f(a).
+   */
+  def unfold[A](o: Option[A])(f: A => Option[A]): Streaming[A] =
+    o match {
+      case None => Empty()
+      case Some(a) => This(a, Always(unfold(f(a))(f)))
+    }
+
+  /**
+   * An empty loop, will wait forever if evaluated.
+   */
+  def godot: Streaming[Nothing] =
```

Comment:
Probably not. We should probably remove this, I was having a bit too much fun! :cat: